Source/JavaScriptCore/runtime/StringPrototype.cpp

```diff
@@ -525,18 +525,36 @@ static ALWAYS_INLINE JSString* replaceUsingRegExpSearchWithCache(VM& vm, JSGloba
     RELEASE_AND_RETURN(scope, jsSpliceSubstringsWithSeparators(globalObject, string, source, sourceRanges.data(), sourceRanges.size(), replacements.data(), replacements.size()));
 }
 
+static ALWAYS_INLINE JSString* tryReplaceOneChar(JSGlobalObject* globalObject, JSString* string, RegExp* regExp, const CallData& callData, const String& replacementString, JSValue replaceValue)
+{
+    if (!string->isRope())
+        return nullptr;
+    if (regExp->global() || !regExp->hasOneCharPattern() || string->length() > 0xFF)
+        return nullptr;
+    if (callData.type != CallData::Type::None || replacementString.find('$') != notFound)
+        return nullptr;
+
+    UChar search = regExp->pattern()[0];
+    if (JSString* result = string->replaceOneChar(globalObject, search, replaceValue.toString(globalObject)))
+        return result;
+    return nullptr;
+}
+
 static ALWAYS_INLINE JSString* replaceUsingRegExpSearch(
     VM& vm, JSGlobalObject* globalObject, JSString* string, JSValue searchValue, const CallData& callData,
     const String& replacementString, JSValue replaceValue)
 {
     auto scope = DECLARE_THROW_SCOPE(vm);
 
+    RegExpObject* regExpObject = jsCast<RegExpObject*>(searchValue);
+    RegExp* regExp = regExpObject->regExp();
+    if (JSString* result = tryReplaceOneChar(globalObject, string, regExp, callData, replacementString, replaceValue))
+        return result;
+
```

Comment:
Discussed offline. Where we should apply this optimization is when using `string.replace(string, string)`, not `string.replace(regexp, string)`.